File: src/client/java/bewis09/bewisclient/hat/HatFeatureRenderer.java
package bewis09.bewisclient.hat;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.model.*;
import net.minecraft.client.render.OverlayTexture;
import net.minecraft.client.render.RenderLayer;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.render.entity.feature.FeatureRenderer;
import net.minecraft.client.render.entity.feature.FeatureRendererContext;
import net.minecraft.client.render.entity.model.EntityModelPartNames;
import net.minecraft.client.render.entity.model.PlayerEntityModel;
import net.minecraft.client.render.entity.state.PlayerEntityRenderState;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.Identifier;

import java.util.Objects;

/**
 * The {@link FeatureRenderer} for the hat
 */
public class HatFeatureRenderer extends FeatureRenderer<PlayerEntityRenderState, PlayerEntityModel> {

    final ModelPart head;
    final ModelPart hat;
    final PlayerEntityModel model;

    public HatFeatureRenderer(FeatureRendererContext<PlayerEntityRenderState, PlayerEntityModel> context) {
        super(context);
        model = context.getModel();
        head = context.getModel().getHead();
        ModelData modelData = new ModelData();
        ModelPartData modelPartData = modelData.getRoot();
        hat = modelPartData.addChild(EntityModelPartNames.HAT, ModelPartBuilder.create().uv(0,0).cuboid(-4,-8,-4,8,8,8,new Dilation(0.51f)),ModelTransform.pivot(0.0F, 0.0F, 0.0F)).createPart(32,16);
    }

    @Override
    public void render(MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, PlayerEntityRenderState state, float limbAngle, float limbDistance) {
        if (!state.invisible && !state.spectator) {
            if (Objects.equals(state.name, MinecraftClient.getInstance().getGameProfile().getName())) {
                {
                    hat.yaw = head.yaw;
                    hat.pitch = head.pitch;
                    hat.roll = head.roll;
                    hat.pivotX = head.pivotX;
                    hat.pivotY = head.pivotY;
                    hat.pivotZ = head.pivotZ;
                }
                hat.render(matrices, vertexConsumers.getBuffer(RenderLayer.getEntityTranslucent(getTexture())), light, OverlayTexture.DEFAULT_UV);
            }
        }
    }

    protected Identifier getTexture() {
        return Hat.current_hat.texture();
    }
}

File: src/client/kotlin/bewis09/bewisclient/cosmetics/feature_renderer/HatFeatureRenderer.kt
package bewis09.bewisclient.cosmetics.feature_renderer

import bewis09.bewisclient.cosmetics.Cosmetics
import net.minecraft.client.model.*
import net.minecraft.client.render.OverlayTexture
import net.minecraft.client.render.RenderLayer
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.entity.feature.FeatureRenderer
import net.minecraft.client.render.entity.feature.FeatureRendererContext
import net.minecraft.client.render.entity.model.EntityModelPartNames
import net.minecraft.client.render.entity.model.PlayerEntityModel
import net.minecraft.client.render.entity.state.PlayerEntityRenderState
import net.minecraft.client.util.math.MatrixStack

/**
 * The [FeatureRenderer] for the hat
 */
class HatFeatureRenderer(context: FeatureRendererContext<PlayerEntityRenderState, PlayerEntityModel?>) : FeatureRenderer<PlayerEntityRenderState, PlayerEntityModel?>(context) {
    val head: ModelPart = context.model!!.getHead()
    val hat: ModelPart

    init {
        val modelData = ModelData()
        val modelPartData = modelData.root
        hat = modelPartData.addChild(EntityModelPartNames.HAT, ModelPartBuilder.create().uv(0, 0).cuboid(-4f, -8f, -4f, 8f, 8f, 8f, Dilation(0.51f)), ModelTransform.pivot(0.0f, 0.0f, 0.0f))
            .createPart(32, 16)
    }

    override fun render(matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, state: PlayerEntityRenderState, limbAngle: Float, limbDistance: Float) {
        if (!state.invisible && !state.spectator) {
            val texture = getTexture(state.name)

            if (texture != null) {
                hat.yaw = head.yaw
                hat.pitch = head.pitch
                hat.roll = head.roll
                hat.pivotX = head.pivotX
                hat.pivotY = head.pivotY
                hat.pivotZ = head.pivotZ

                hat.render(matrices, vertexConsumers.getBuffer(RenderLayer.getEntityTranslucent(texture)), light, OverlayTexture.DEFAULT_UV)
            }
        }
    }

    fun getTexture(name: String) = Cosmetics.hats.getTexture(name)
}